saleor/product/utils/variant_prices.py

```diff
@@ -5,52 +5,63 @@
 from django.db.models.query_utils import Q
 from prices import Money
 
-from ...discount.utils import fetch_active_discounts
-from ..models import Product
+from ...discount.utils import calculate_discounted_price, fetch_active_discounts
+from ..models import Product, ProductChannelListing, ProductVariantChannelListing
 
 
-def _get_product_minimal_variant_price(product, discounts) -> Optional[Money]:
+def _get_product_discounted_price(
+    product_channel_listing, discounts
+) -> Optional[Money]:
     # Start with the product's price as the minimal one
     minimal_variant_price = None
-    for variant in product.variants.all():
-        variant_price = variant.get_price(discounts=discounts)
+    variants_channel_listing = ProductVariantChannelListing.objects.filter(
+        variant__product_id=product_channel_listing.product_id,
+        channel_id=product_channel_listing.channel_id,
+    )
+    product = product_channel_listing.product
+    collections = (product.collections.all(),)
+    for variant_channel_listing in variants_channel_listing:
+        variant_price = calculate_discounted_price(
+            product=product,
+            price=variant_channel_listing.price,
+            collections=collections,
+            discounts=discounts,
+        )
         if minimal_variant_price is None:
             minimal_variant_price = variant_price
         else:
             minimal_variant_price = min(minimal_variant_price, variant_price)
     return minimal_variant_price
 
 
-def update_product_minimal_variant_price(product, discounts=None, save=True):
+def update_product_discounted_price(product, discounts=None):
     if discounts is None:
         discounts = fetch_active_discounts()
-    minimal_variant_price = _get_product_minimal_variant_price(product, discounts)
-    if product.minimal_variant_price != minimal_variant_price:
-        product.minimal_variant_price_amount = minimal_variant_price.amount
-        if save:
-            product.save(update_fields=["minimal_variant_price_amount", "updated_at"])
-    return product
+    product_channel_listings = product.channel_listing.all()
+    changed_products_channels_to_update = []
+    for product_channel_listing in product_channel_listings:
+        product_discounted_price = _get_product_discounted_price(
+            product_channel_listing, discounts
+        )
```

Comment:
`_get_product_discounted_price` makes a database query with `ProductVariantChannelListing.objects.filter()` so we will have duplicated queries in this loop. Maybe we can improve the performance here?